ENH BaseLabelPropagation to accept sparse matrices

cozek opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->
Fixes #17085
Closes #17384
#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Also see stale PR on the same issue https://github.com/scikit-learn/scikit-learn/pull/17384

#### What does this implement/fix? Explain your changes.
At present `LabelPropagation` and `LabelSpreading` throw  `TypeError` for sparse matrices.

As per my investigation this error arises because the function `_validate_data` function in `BaseLabelPropagation.fit` method hasn't been called with the `accept_sparse` parameter.

This PR fixes the issue by passing the same parameters to `_validate_data` as passed in `predict_proba` method of the 
`BaseLabelPropagation` class.

In addition, the `test_fit_transduction` func  in `test_label_propagation.py` has been updated to test for sparse metrices
using `pytest.mark.parametrize`.

#### Any other comments?
Most of tests in `test_label_propagation.py` can be parametrized. I have parametrized the func `test_fit_transduction`
in the following format

Firstly the global variable `ESTIMATORS` can be rewritten as follows for (what i believe to be more pythonic) to reduce redundancy.
```python
ESTIMATORS = [
    (label_propagation.LabelPropagation, {'kernel': 'rbf'}),
    (label_propagation.LabelPropagation, {'kernel': 'knn', 'n_neighbors': 2}),
    (label_propagation.LabelPropagation, {
        'kernel': lambda x, y: rbf_kernel(x, y, gamma=20)
    }),
    (label_propagation.LabelSpreading, {'kernel': 'rbf'}),
    (label_propagation.LabelSpreading, {'kernel': 'knn', 'n_neighbors': 2}),
    (label_propagation.LabelSpreading, {
        'kernel': lambda x, y: rbf_kernel(x, y, gamma=20)
    }),
]

# TO
ESTIMATOR_CLS = [label_propagation.LabelPropagation,
                 label_propagation.LabelSpreading]


PARAMETERS = [{'kernel': 'rbf'}, {'kernel': 'knn', 'n_neighbors': 2}, {
        'kernel': lambda x, y: rbf_kernel(x, y, gamma=20)
    }]
```

This breakdown to two globals allows for parametrization using the decorators
```python
@pytest.mark.parametrize("estimator", ESTIMATOR_CLS)
@pytest.mark.parametrize("parameters", PARAMETERS)
def func(estimator, parameters):
    pass
``` 
This is what I have done with the test that I have modified. This same thing can be done to many of the tests. 
And then we can remove the `ESTIMATORS` global variable completely.
But I haven't touched them since this is a deviation from the issue. 
Perhaps this can be done in a separate PR. (or this one)

```python
ESTIMATOR_CLS = [label_propagation.LabelPropagation,
                 label_propagation.LabelSpreading]


PARAMETERS = [{'kernel': 'rbf'}, {'kernel': 'knn', 'n_neighbors': 2}, {
        'kernel': lambda x, y: rbf_kernel(x, y, gamma=20)
    }]


MATRIX_TYPES = (sp.csc_matrix, sp.csr_matrix, sp.coo_matrix, sp.dok_matrix,
                sp.bsr_matrix, sp.lil_matrix, sp.dia_matrix, np.array)


@pytest.mark.parametrize("matrix_type", MATRIX_TYPES)
@pytest.mark.parametrize("estimator", ESTIMATOR_CLS)
@pytest.mark.parametrize("parameters", PARAMETERS)
def test_fit_transduction(matrix_type, estimator, parameters):
    samples = [[1., 0.], [0., 2.], [1., 3.]]
    labels = [0, 1, -1]
    clf = estimator(**parameters).fit(matrix_type(samples), labels)
    assert clf.transduction_[2] == 1
```

Also, we can take the following as  global and use parametrize
```
ALPHA = [0.1, 0.3, 0.5, 0.7, 0.9]
SAMPLE = [[1., 0.], [0., 1.], [1., 1.]]
LABELS = [0, 1, -1]
```
Just let me know how I should proceed regarding this. Thanks for reading! :)

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

cozek committed:
initial commit with fixes

cozek committed:
fix flake8 problems

cozek committed:
add to changelog

cozek committed:
fix changelog edit

cozek commented:
@cmarmo if possible, I would appreciate  a "waiting for reviewer" label. :)

cmarmo commented:
@cozek , thanks for the ping! :)
I'm not sure your pull request will make it for 0.24.2 (in preparation right now).
Perhaps you will be asked to move the changelog entry in 1.0.

jjerphan commented on sklearn/semi_supervised/tests/test_label_propagation.py:
```diff
@@ -26,13 +27,27 @@
 ]
 
+ESTIMATOR_CLS = [label_propagation.LabelPropagation,
+                 label_propagation.LabelSpreading]
 
-def test_fit_transduction():
+
+PARAMETERS = [{'kernel': 'rbf'}, {'kernel': 'knn', 'n_neighbors': 2}, {
+        'kernel': lambda x, y: rbf_kernel(x, y, gamma=20)
+    }]
+
+
+MATRIX_TYPES = (sp.csc_matrix, sp.csr_matrix, sp.coo_matrix, sp.dok_matrix,
+                sp.bsr_matrix, sp.lil_matrix, sp.dia_matrix, np.array)
```
I agree with your point and this setup: using test parametrization definitely would be better. :+1: 

You can proceed in two steps in different commits (at least two but not necessarily) to:
 - adapt the other tests to assert the correct behavior of `fit` on sparse matrices
 - if all the tests pass, adapt the test to use parametrisation as you started to do with `test_fit_transduction`
 
If we judge that it would be better having a dedicated PR for this second step, we can discard them here but cherry-pick them on another PR branch. :slightly_smiling_face:

cozek replied:
I did this due the reviewer comments in the stalled PR that this PR takes-over
https://github.com/scikit-learn/scikit-learn/pull/17384#discussion_r454514567

I agree that we should have a dedicated PR instead. I will revert these edits and only focus only on the main the tests/funcs.

jjerphan commented on sklearn/semi_supervised/tests/test_label_propagation.py:
```diff
@@ -26,13 +27,27 @@
     samples = [[1., 0.], [0., 2.], [1., 3.]]
     labels = [0, 1, -1]
-    for estimator, parameters in ESTIMATORS:
-        clf = estimator(**parameters).fit(samples, labels)
-        assert clf.transduction_[2] == 1
+    clf = estimator(**parameters).fit(matrix_type(samples), labels)
```
I can suggest on this patch, but it's easier to see the matrix type this way, I think.
```python 
    samples = matrix_type([[1., 0.], [0., 2.], [1., 3.]])
    labels = [0, 1, -1]
    clf = estimator(**parameters).fit(samples, labels)
```

jjerphan commented on doc/whats_new/v0.24.rst:
```diff
@@ -12,6 +12,12 @@ Version 0.24.2
 Changelog
 ---------
 
+:mod:`sklearn.semi_supervised`
+..............................
+
+- |Fix| :meth:`BaseLabelPropagation.fit` does not accept sparse matrix.
+  :pr:`19664` by :user:`Kaushik Amar Das <cozek>`.
+
```
As @cmarmo indicated, this has to be moved to `doc/whats_new/v1.0.rst`.
```suggestion
:mod:`sklearn.semi_supervised`
..............................

- |Fix| :meth:`BaseLabelPropagation.fit` now accepts sparse matrices.
  :pr:`19664` by :user:`Kaushik Amar Das <cozek>`.

```

jjerphan commented on sklearn/semi_supervised/_label_propagation.py:
```diff
@@ -226,7 +226,10 @@ def fit(self, X, y):
         self : object
         """
-        X, y = self._validate_data(X, y)
+        X, y = self._validate_data(
+            X, y, accept_sparse=['csc', 'csr',
+                                 'coo', 'dok', 'bsr', 'lil', 'dia'],
+            reset=True)
```
```suggestion
        X, y = self._validate_data(
            X, y, accept_sparse=['csc', 'csr', 'coo', 'dok',
                                 'bsr', 'lil', 'dia'],
            reset=True)
```

jjerphan reviewed (changes requested):
Thanks @cozek: this contribution is indeed useful!

Here are a few comments.

cozek committed:
Merge branch 'main' into fix-labelprop

jjerphan commented:
Hi @cozek, do you need any help with this PR? :slightly_smiling_face:

cozek commented:
> Hi @cozek, do you need any help with this PR? 🙂
I will work on this over the week and should have updated it by next week following your suggestions. Sorry about the delay.

jjerphan commented:
> Sorry about the delay.

No problems: it's fine. :)

cozek committed:
Merge branch 'scikit-learn:main' into fix-labelprop

cozek committed:
update formatting

Commit:
update formatting

cozek committed:
Merge branch 'fix-labelprop' of https://github.com/cozek/scikit-learn into fix-labelprop

cozek committed:
update test_fit_transduction

cozek committed:
updating whatsnew

cozek committed:
test for sparse in test_distribution

cozek committed:
test for sparse in test_predict

cozek committed:
test for sparse in test_predict_proba

cozek committed:
test for sparse in test_label_spreading_closed_form

cozek committed:
test for sparse in test_label_propagation_closed_form

cozek committed:
test for sparse in test_valid_alpha

cozek committed:
test for sparse in test_convergence_speed

cozek committed:
test for sparse in test_convergence_warning

cozek committed:
test for sparse in test_label_propagation_non_zero_normalizer

cozek committed:
test for sparse in test_predict_sparse_callable_kernel

cozek committed:
fix flake8 linting

cozek committed:
removing ESTIMATORS and using parametrize instead

jjerphan commented on sklearn/semi_supervised/tests/test_label_propagation.py:
```diff
@@ -13,63 +14,70 @@
+    samples = matrix_type([[1., 0.], [0., 1.], [1., 1.]])
     labels = [0, 1, -1]
-    for estimator, parameters in ESTIMATORS:
-        clf = estimator(**parameters).fit(samples, labels)
-        if parameters['kernel'] == 'knn':
-            continue    # unstable test; changes in k-NN ordering break it
-            assert_array_almost_equal(clf.predict_proba([[1., 0.0]]),
-                                      np.array([[1., 0.]]), 2)
-        else:
-            assert_array_almost_equal(np.asarray(clf.label_distributions_[2]),
-                                      np.array([.5, .5]), 2)
```
We probably want to maintain both cases here for this test.

cozek replied:
Parameterisation removes the looping, so `continue` doesn't work. Plus, that test under the if seems unreachable to me even in the original function.

doing the following fails 16 tests.
```python
if parameters['kernel'] != 'knn':
        # unstable test; changes in k-NN ordering break it
        assert_array_almost_equal(clf.predict_proba([[1., 0.0]]),
                                  np.array([[1., 0.]]), 2)
        assert_array_almost_equal(np.asarray(clf.label_distributions_[2]),
                                  np.array([.5, .5]), 2)
```
Also, `dok` fails some convergence tests in `test_convergence_warning` and  `test_label_propagation_non_zero_normalizer`

I would appreciate some suggestions regarding how to proceed regarding the above issues. Thanks!

jjerphan replied:
Thank for the explanation, it makes sense.

Probably you might prefer using [`pytest.skip`](https://docs.pytest.org/en/reorganize-docs/new-docs/user/skipping.html) on such tests case.

> `dok` fails some convergence tests in `test_convergence_warning` and `test_label_propagation_non_zero_normalizer`

Do you have pointers to failing tests log?

jjerphan replied:
I am currently busy with others tasks and inspecting this bug will take a bit of a time as I am not familiar with this part of the code base yet. Thanks for reporting this and for your patience. 🙂

cozek replied:
I understand. No issues. 😄

glemaitre committed:
Merge remote-tracking branch 'origin/main' into pr/cozek/19664-1

glemaitre commented on sklearn/semi_supervised/tests/test_label_propagation.py:
```diff
@@ -13,67 +14,78 @@
-        {"kernel": lambda x, y: rbf_kernel(x, y, gamma=20)},
-    ),
+
+ESTIMATOR_CLS = [label_propagation.LabelPropagation, label_propagation.LabelSpreading]
```
Actually, since we are always using the same model, I would prefer to use a `pytest.fixture` to generate once for all the different estimators with different parameters.

glemaitre commented on sklearn/semi_supervised/tests/test_label_propagation.py:
```diff
@@ -13,67 +14,78 @@
-    for estimator, parameters in ESTIMATORS:
-        clf = estimator(**parameters).fit(samples, labels)
-        assert clf.transduction_[2] == 1
+MATRIX_TYPES = (
```
I would only try `csr` `csc` in the test.

glemaitre commented on sklearn/semi_supervised/tests/test_label_propagation.py:
```diff
@@ -13,67 +14,78 @@
+@pytest.mark.parametrize("estimator", ESTIMATOR_CLS)
+@pytest.mark.parametrize("parameters", PARAMETERS)
+def test_fit_transduction(matrix_type, estimator, parameters):
+    samples = matrix_type([[1.0, 0.0], [0.0, 2.0], [1.0, 3.0]])
```
instead of `matrix_type`, we have a utility in testing to make the conversion. You can look at `sklearn.utils._testing._convert_container`. Like this, you can try array, and the two types of sparse matrix.

glemaitre reviewed (commented):
I would additionally add a test to check that passing a sparse matrix or a dense matrix will lead to the same result

cmarmo commented:
Hi @cozek , are you still interested in finishing this PR? Do you mind addressing the comments about the tests? Thanks.

cozek commented:
> Hi @cozek , are you still interested in finishing this PR? Do you mind addressing the comments about the tests? Thanks.

Yes. I will get to it this weekend.

cozek committed:
reviewer changes

cozek committed:
Merge branch 'main' into fix-labelprop

cozek committed:
Merge branch 'fix-labelprop' of https://github.com/cozek/scikit-learn into fix-labelprop

glemaitre commented on doc/whats_new/v1.0.rst:
```diff
@@ -274,6 +274,12 @@ Changelog
+:mod:`sklearn.semi_supervised`
+..............................
+
+- |Fix| :meth:`BaseLabelPropagation.fit` now accepts sparse matrices.
```
Actually, this is an enhancement and not a fix.
You can use `|Enhancement|`. This entry should be moved in the file `v1.1.rst` now.

glemaitre commented on sklearn/semi_supervised/_label_propagation.py:
```diff
@@ -246,7 +246,12 @@ def fit(self, X, y):
             Returns the instance itself.
         """
-        X, y = self._validate_data(X, y)
+        X, y = self._validate_data(
```
Actually, I see that `LabelPropagation` is redefining the docstring as well. So we should propagate the change in the docstring there.

glemaitre commented on sklearn/semi_supervised/tests/test_label_propagation.py:
```diff
@@ -12,68 +13,85 @@
+
+
+@pytest.fixture
+def estimators():
```
I did not give good advice in my previous review. Indeed, we don't need a `fixture` but just a function. This function could generate the different candidates to be tested and be used inside a `parametrize` function.

I am putting the diff of such modification here:

```diff
diff --git a/sklearn/semi_supervised/tests/test_label_propagation.py b/sklearn/semi_supervised/tests/test_label_propagation.py
index 7cad54adf7..490bae22f3 100644
--- a/sklearn/semi_supervised/tests/test_label_propagation.py
+++ b/sklearn/semi_supervised/tests/test_label_propagation.py
@@ -1,4 +1,5 @@
 """ test the label propagation module """
+import itertools
 
 import numpy as np
 import pytest
@@ -15,76 +16,68 @@ from numpy.testing import assert_array_almost_equal
 from numpy.testing import assert_array_equal
 from sklearn.utils._testing import _convert_container
 
-ESTIMATOR_CLS = [label_propagation.LabelPropagation, label_propagation.LabelSpreading]
-
-
-PARAMETERS = [
-    {"kernel": "rbf"},
-    {"kernel": "knn", "n_neighbors": 2},
-    {"kernel": lambda x, y: rbf_kernel(x, y, gamma=20)},
-]
-
-
 CONSTRUCTOR_TYPES = ("array", "sparse_csr", "sparse_csc")
 
 
-@pytest.fixture
-def estimators():
-    estimators_ = []
-    for estimator in ESTIMATOR_CLS:
-        for params in PARAMETERS:
-            estimators_.append(estimator(**params))
-    return estimators_
+def label_estimator():
+    LabelEstimator = [label_propagation.LabelPropagation, label_propagation.LabelSpreading]
+    parama_grid = [
+        {"kernel": "rbf"},
+        {"kernel": "knn", "n_neighbors": 2},
+        {"kernel": lambda x, y: rbf_kernel(x, y, gamma=20)},
+    ]
+    for Estimator, params in itertools.product(LabelEstimator, parama_grid):
+        yield Estimator(**params)
 
 
 @pytest.mark.parametrize("constructor_type", CONSTRUCTOR_TYPES)
-def test_fit_transduction(constructor_type, estimators):
-    for estimator in estimators:
-        samples = _convert_container(
-            [[1.0, 0.0], [0.0, 2.0], [1.0, 3.0]], constructor_type
-        )
-        labels = [0, 1, -1]
-        clf = estimator.fit(samples, labels)
-        assert clf.transduction_[2] == 1
+@pytest.mark.parametrize("estimator", label_estimator())
+def test_fit_transduction(constructor_type, estimator):
+    samples = _convert_container(
+        [[1.0, 0.0], [0.0, 2.0], [1.0, 3.0]], constructor_type
+    )
+    labels = [0, 1, -1]
+    clf = estimator.fit(samples, labels)
+    assert clf.transduction_[2] == 1
 
 
 @pytest.mark.parametrize("constructor_type", CONSTRUCTOR_TYPES)
-def test_distribution(constructor_type, estimators):
-    for estimator in estimators:
-        samples = _convert_container(
-            [[1.0, 0.0], [0.0, 1.0], [1.0, 1.0]], constructor_type
+@pytest.mark.parametrize("estimator", label_estimator())
+def test_distribution(constructor_type, estimator):
+    samples = _convert_container(
+        [[1.0, 0.0], [0.0, 1.0], [1.0, 1.0]], constructor_type
+    )
+    labels = [0, 1, -1]
+    clf = estimator.fit(samples, labels)
+    if estimator.kernel != "knn":
+        # unstable test; changes in k-NN ordering break it
+        assert_array_almost_equal(
+            np.asarray(clf.label_distributions_[2]), np.array([0.5, 0.5]), 2
         )
-        labels = [0, 1, -1]
-        clf = estimator.fit(samples, labels)
-        if estimator.kernel != "knn":
-            # unstable test; changes in k-NN ordering break it
-            assert_array_almost_equal(
-                np.asarray(clf.label_distributions_[2]), np.array([0.5, 0.5]), 2
-            )
 
 
 @pytest.mark.parametrize("constructor_type", CONSTRUCTOR_TYPES)
-def test_predict(constructor_type, estimators):
-    for estimator in estimators:
-        samples = _convert_container(
-            [[1.0, 0.0], [0.0, 2.0], [1.0, 3.0]], constructor_type
-        )
-        labels = [0, 1, -1]
-        clf = estimator.fit(samples, labels)
-        assert_array_equal(clf.predict([[0.5, 2.5]]), np.array([1]))
+@pytest.mark.parametrize("estimator", label_estimator())
+def test_predict(constructor_type, estimator):
+    samples = _convert_container(
+        [[1.0, 0.0], [0.0, 2.0], [1.0, 3.0]], constructor_type
+    )
+    labels = [0, 1, -1]
+    clf = estimator.fit(samples, labels)
+    assert_array_equal(clf.predict([[0.5, 2.5]]), np.array([1]))
 
 
 @pytest.mark.parametrize("constructor_type", CONSTRUCTOR_TYPES)
-def test_predict_proba(constructor_type, estimators):
-    for estimator in estimators:
-        samples = _convert_container(
-            [[1.0, 0.0], [0.0, 1.0], [1.0, 2.5]], constructor_type
-        )
-        labels = [0, 1, -1]
-        clf = estimator.fit(samples, labels)
-        assert_array_almost_equal(
-            clf.predict_proba([[1.0, 1.0]]), np.array([[0.5, 0.5]])
-        )
+@pytest.mark.parametrize("estimator", label_estimator())
+def test_predict_proba(constructor_type, estimator):
+    samples = _convert_container(
+        [[1.0, 0.0], [0.0, 1.0], [1.0, 2.5]], constructor_type
+    )
+    labels = [0, 1, -1]
+    clf = estimator.fit(samples, labels)
+    assert_array_almost_equal(
+        clf.predict_proba([[1.0, 1.0]]), np.array([[0.5, 0.5]])
+    )
 
 
 @pytest.mark.parametrize("constructor_type", CONSTRUCTOR_TYPES)
@@ -187,7 +180,10 @@ def test_convergence_warning(constructor_type):
 
 
 @pytest.mark.parametrize("constructor_type", CONSTRUCTOR_TYPES)
-@pytest.mark.parametrize("LabelPropagationCls", ESTIMATOR_CLS)
+@pytest.mark.parametrize(
+    "LabelPropagationCls",
+    [label_propagation.LabelPropagation, label_propagation.LabelSpreading],
+)
 def test_label_propagation_non_zero_normalizer(LabelPropagationCls, constructor_type):
     # check that we don't divide by zero in case of null normalizer
     # non-regression test for
```

glemaitre reviewed (commented):
Otherwise, I am fine with the rest.

cozek committed:
Merge branch 'main' into fix-labelprop

cozek committed:
adding a function to generate different testing candidates and using it with parametrize

cozek committed:
Merge branch 'main' into fix-labelprop

cozek committed:
adding changes to whats_new

cozek committed:
updating docstrings for LabelPropagation and BaseLabelPropagation

cmarmo commented:
Thanks @cozek . Note that the failing check is unrelated with your modifications.

cmarmo commented:
Hi @cozek, apparently finding a reviewer takes some time.... sorry for that... do you mind synchronizing with upstream and fixing conflicts? This will also make easier the work of the reviewers. Thank you very much for your patience!

thomasjpfan commented on doc/whats_new/v1.0.rst:
```diff
@@ -280,6 +280,9 @@ Changelog
     :pr:`123456` by :user:`Joe Bloggs <joeongithub>`.
     where 123456 is the *pull request* number, not the issue number.
 
+:mod:`sklearn.semi_supervised`
```
May we remove this header?

thomasjpfan commented on doc/whats_new/v1.1.rst:
```diff
@@ -70,6 +70,12 @@ Changelog
   error message when setting invalid hyper-parameters with `set_params`.
   :pr:`21542` by :user:`Olivier Grisel <ogrisel>`.
 
+:mod:`sklearn.semi_supervised`
```
This needs to be moved to 1.2 now.

thomasjpfan commented on sklearn/semi_supervised/_label_propagation.py:
```diff
@@ -246,7 +246,12 @@ def fit(self, X, y):
+        X, y = self._validate_data(
+            X,
+            y,
+            accept_sparse=["csc", "csr", "coo", "dok", "bsr", "lil", "dia"],
```
I'll prefer to start with "csc" and "csr" and add others if they are requested.

```suggestion
            accept_sparse=["csr", "csc"],
```

These are also the ones that are tested.

thomasjpfan commented on sklearn/semi_supervised/tests/test_label_propagation.py:
```diff
@@ -12,68 +14,70 @@
+CONSTRUCTOR_TYPES = ("array", "sparse_csr", "sparse_csc")
+
+
+def label_estimator():
```
How much does this increase the runtime of running this test file? In other words, how long does the following command take on `main` vs this PR?

```bash
pytest sklearn/semi_supervised/tests/test_label_propagation.py
```

cozek replied:
On my machine, `main` runs in ~0.55s, this pr runs in ~0.60s.

thomasjpfan reviewed (commented):
Thank you for the PR! There are some merge conflicts to resolve.

cozek commented:
Hi. Thanks for the review @thomasjpfan  . Will work on these as soon as I can.

cozek committed:
merging with main

jjerphan commented:
Hi @cozek, do you still have time to work on this PR?

cozek commented:
Working on it today.

cozek committed:
adding changes

cozek committed:
updating docs to reflect changes

cozek committed:
complying with black and autoflake

cozek committed:
removing unnecessary lines

glemaitre committed:
Merge branch 'main' into fix-labelprop

jjerphan commented on sklearn/semi_supervised/tests/test_label_propagation.py:
```diff
@@ -1,5 +1,4 @@
 """ test the label propagation module """
-
```
This should must be reverted.

cozek replied:
Should I add the blank line or remove following changes in `test_convergene_speed` ?

```
from sklearn.utils._testing import _convert_container
CONSTRUCTOR_TYPES = ("array", "sparse_csr", "sparse_csc")

@pytest.mark.parametrize("constructor_type", CONSTRUCTOR_TYPES)
def test_convergence_speed(constructor_type):
    # This is a non-regression test for #5774
    X = _convert_container([[1.0, 0.0], [0.0, 1.0], [1.0, 2.5]], constructor_type)
    y = np.array([0, 1, -1])
    mdl = label_propagation.LabelSpreading(kernel="rbf", max_iter=5000)
    mdl.fit(X, y)

    # this should converge quickly:
    assert mdl.n_iter_ < 10
    assert_array_equal(mdl.predict(X), [0, 1, 1])
```

jjerphan commented on sklearn/semi_supervised/_label_propagation.py:
```diff
@@ -255,8 +255,12 @@ def fit(self, X, y):
+        X, y = self._validate_data(
+            X,
+            y,
+            accept_sparse=["csr", "csc"],
```
`BaseEstimator._validate_data` forwards this parameter to `sklearn.utils.check_array`. This function has parameter called `accept_large_sparse` which we might want to specify the value of for this call for `BaseLabelPropagation`:

https://github.com/scikit-learn/scikit-learn/blob/550fa9c28c7193095487563ebc071271427b2bc3/sklearn/utils/validation.py#L664-L669

Could you:
 - add a test which specifically passes CSR and CSC matrices with `np.int64`-typed instead of `np.int32`-typed (the default) `indices` and `indptr` to assert the behavior of classes extending `BaseLabelPropagation` (i.e. assert if those implementation accept such inputs or if they do not)
 - specify the value of `accept_large_sparse` according to this behavior?
 
To get to `np.int64`-typed `indices` and `indptr` from `np.int32`-typed `indices` and `indptr`, this should suffice:

```python
# CSR chosen without loss of generality
X_csr.indices = X_csr.indices.astype(np.int64, copy=False) 
X_csr.indptr = X_csr.indptr.astype(np.int64, copy=False)
```

jjerphan replied:
Yes! At least, let's start with this one.

jjerphan reviewed (commented):
Thanks for the update, @cozek!

cozek committed:
Merge branch 'main' into fix-labelprop

cozek committed:
Merge branch 'fix-labelprop' of https://github.com/cozek/scikit-learn into fix-labelprop

cozek committed:
reverting changes

cozek committed:
reverting changes

cozek committed:
adding tests for sparse int32 and int64

cozek committed:
Merge branch 'main' into fix-labelprop

jjerphan commented on sklearn/semi_supervised/_label_propagation.py:
```diff
@@ -255,8 +255,12 @@ def fit(self, X, y):
         """
-        self._validate_params()
-        X, y = self._validate_data(X, y)
+        X, y = self._validate_data(
```
I think this line was removed when updating the PR. It will fix failing tests on the CI.
```suggestion
        self._validate_params()

        X, y = self._validate_data(
```

cozek replied:
done. 

In this pr, the `test_label_propagation.py` takes around 0.75s vs ~0.55s of `main`. (in my laptop)

jjerphan reviewed (commented):
I think we are nearly there.

cozek committed:
fixing docs and adding params validation

cozek committed:
Merge branch 'main' into fix-labelprop

thomasjpfan commented on sklearn/semi_supervised/tests/test_label_propagation.py:
```diff
@@ -122,9 +125,23 @@ def test_label_propagation_closed_form(global_dtype):
+@pytest.mark.parametrize("accepted_sparse_type", ["sparse_csr", "sparse_csc"])
+@pytest.mark.parametrize("dtype", [np.int32, np.int64])
+def test_input_types(accepted_sparse_type, dtype):
+    # This is non-regression test for #19664
```
I think #17085 was the original issue:

```suggestion
    # This is non-regression test for #17085
```

thomasjpfan commented on sklearn/semi_supervised/tests/test_label_propagation.py:
```diff
@@ -122,9 +125,23 @@ def test_label_propagation_closed_form(global_dtype):
 
-def test_convergence_speed():
+@pytest.mark.parametrize("accepted_sparse_type", ["sparse_csr", "sparse_csc"])
+@pytest.mark.parametrize("dtype", [np.int32, np.int64])
```
Can we change this name to `index_dtype` because it is changing the dtype for the `indices` and the `indptr`?

```suggestion
@pytest.mark.parametrize("index_dtype", [np.int32, np.int64])
```

The rest of the function will need to be updated.

thomasjpfan reviewed (commented):
Thanks for the update! Overall looks good.

cozek committed:
resolving conflict

cozek committed:
reviewer changes

jjerphan commented on sklearn/semi_supervised/tests/test_label_propagation.py:
```diff
@@ -122,9 +125,23 @@ def test_label_propagation_closed_form(global_dtype):
+    X.indices = X.indices.astype(index_dtype, copy=False)
+    X.indptr = X.indptr.astype(index_dtype, copy=False)
+    y = np.array([0, 1, -1])
+    mdl = label_propagation.LabelSpreading(kernel="rbf", max_iter=5000)
```
I think the behavior of `LabelPropagation` is the only thing left to test.

cozek replied:
I think, replacing that test with the one below covers it all. What do you think?
Its based on `test_predict`.

```py
@pytest.mark.parametrize("accepted_sparse_type", ["sparse_csr", "sparse_csc"])
@pytest.mark.parametrize("index_dtype", [np.int32, np.int64])
@pytest.mark.parametrize("Estimator, parameters", ESTIMATORS)
def test_sparse_input_types(accepted_sparse_type, index_dtype, Estimator, parameters):
    # This is non-regression test for #17085
    X = _convert_container([[1.0, 0.0], [0.0, 2.0], [1.0, 3.0]], accepted_sparse_type)
    X.indices = X.indices.astype(index_dtype, copy=False)
    X.indptr = X.indptr.astype(index_dtype, copy=False)
    labels = [0, 1, -1]
    clf = Estimator(**parameters).fit(X, labels)
    assert_array_equal(clf.predict([[0.5, 2.5]]), np.array([1]))
```

jjerphan replied:
This test parametrisation looks good to me. :+1:

cozek committed:
improving tests

cozek committed:
Merge branch 'main' into fix-labelprop

cozek committed:
Merge branch 'main' into fix-labelprop

cozek commented:
@jjerphan @thomasjpfan  Hi, would appreciate your feedback.

jjerphan commented on sklearn/semi_supervised/tests/test_label_propagation.py:
```diff
@@ -122,9 +125,23 @@ def test_label_propagation_closed_form(global_dtype):
 
-def test_convergence_speed():
+@pytest.mark.parametrize("accepted_sparse_type", ["sparse_csr", "sparse_csc"])
+@pytest.mark.parametrize("index_dtype", [np.int32, np.int64])
+@pytest.mark.parametrize("Estimator, parameters", ESTIMATORS)
+def test_sparse_input_types(accepted_sparse_type, index_dtype, Estimator, parameters):
+    # This is non-regression test for #17085
+    X = _convert_container([[1.0, 0.0], [0.0, 2.0], [1.0, 3.0]], accepted_sparse_type)
```
I think we can parametrize this test on `dtype` as well.

```suggestion
@pytest.mark.parametrize("index_dtype", [np.int32, np.int64])
@pytest.mark.parametrize("dtype", [np.float32, np.float64])
@pytest.mark.parametrize("Estimator, parameters", ESTIMATORS)
def test_sparse_input_types(accepted_sparse_type, index_dtype, dtype, Estimator, parameters):
    # This is non-regression test for #17085
    X = _convert_container([[1.0, 0.0], [0.0, 2.0], [1.0, 3.0]], accepted_sparse_type)
    X.data = X.data.astype(dtype, copy=False)
```

jjerphan reviewed (approved):
LGTM. Just one last comment.

cozek committed:
Merge branch 'main' into fix-labelprop

cozek committed:
Parametrizing dtype

jjerphan commented:
@thomasjpfan: is this PR mergeable to you?

cozek commented:
@thomasjpfan Ping. (I hope you don't mind)

cozek committed:
Merge branch 'main' into fix-labelprop

thomasjpfan reviewed (approved):
Thank you for the update! LGTM
